frontends/bmd/src/pages/accessible_controller_diagnostic_screen.tsx

```diff
@@ -0,0 +1,207 @@
+  }, [buttonKey, buttonName, onSuccess]);
+
+  return (
+    <StepContainer>
+      <div>
+        <h1>Press the {buttonName.toLowerCase()} button.</h1>
+        <Button
+          onPress={() => onFailure(`${buttonName} button is not working.`)}
+        >
+          {buttonName} Button is Not Working
+        </Button>
+      </div>
+      <img src="/images/controller-up-arrow.png" alt="up" />
+    </StepContainer>
+  );
+}
+
+interface AccessibleControllerSoundDiagnosticProps {
+  screenReader: ScreenReader;
+  onSuccess: () => void;
+  onFailure: (message: string) => void;
+}
+
+function AccessibleControllerSoundDiagnostic({
+  screenReader,
+  onSuccess,
+  onFailure,
+}: AccessibleControllerSoundDiagnosticProps) {
+  const [hasPlayedAudio, setHasPlayedAudio] = useState(false);
+
+  useEffect(() => {
+    async function handleKeyDown(event: KeyboardEvent) {
+      event.stopPropagation();
+      if (event.key === 'ArrowRight') {
+        const wasMuted = screenReader.isMuted();
+        screenReader.unmute();
+        await screenReader.speak(
+          'Press the select button to confirm sound is working.'
+        );
+        screenReader.toggleMuted(wasMuted);
+        setHasPlayedAudio(true);
+      }
+      if (event.key === 'Enter' && hasPlayedAudio) {
+        onSuccess();
+      }
+    }
+    document.addEventListener('keydown', handleKeyDown, { capture: true });
+    return () =>
+      document.removeEventListener('keydown', handleKeyDown, { capture: true });
+  }, [screenReader, onSuccess, hasPlayedAudio]);
+
+  return (
+    <StepContainer>
+      <div>
+        <h1>Confirm sound is working.</h1>
+        <ol>
+          <li>Plug in headphones.</li>
+          <li>Press the right button to play sound.</li>
+          <li>Press the select button to confirm sound is working.</li>
+        </ol>
+        <Button onPress={() => onFailure('Sound is not working.')}>
+          Sound is Not Working
+        </Button>
+      </div>
+      <img src="/images/controller-up-arrow.png" alt="up" />
+    </StepContainer>
+  );
+}
+
+export interface AccessibleControllerDiagnosticProps {
+  onComplete: (results: AccessibleControllerDiagnosticResults) => void;
+  onCancel: () => void;
+  screenReader: ScreenReader;
+}
+
+export function AccessibleControllerDiagnostic({
+  onComplete,
+  onCancel,
+  screenReader,
+}: AccessibleControllerDiagnosticProps): JSX.Element {
+  const [step, setStep] = useState(0);
+
+  function passTest() {
+    onComplete({ passed: true, completedAt: DateTime.now() });
+  }
+  function failTest(message: string) {
+    onComplete({ passed: false, completedAt: DateTime.now(), message });
+  }
+
+  function nextStep() {
+    setStep((previousStep) => previousStep + 1);
+  }
+
+  const buttons = [
+    ['Up', 'ArrowUp'],
+    ['Down', 'ArrowDown'],
+    ['Left', 'ArrowLeft'],
+    ['Right', 'ArrowRight'],
+    ['Select', 'Enter'],
+  ];
+  const steps = [
+    ...buttons.map(([buttonName, buttonKey]) => (
```

Comment:
Why not just build the one we need for the current step?

Reply:
the sound diagnostic is also a step though, so it should only be rendered when `step` is 5

Reply:
Oh, whoops. I misread the code. I might try to have a data structure that's just a little more info than the `buttons` one is that says what type it is etc, but I realize that's also introducing similar complexity in another direction. So this is fine.